gio-subclass/src/action.rs

```diff
@@ -0,0 +1,220 @@
+// This file was generated by gir (https://github.com/gtk-rs/gir @ 5b90945+)
+// from gir-files (https://github.com/gtk-rs/gir-files @ b215ee8+)
+// DO NOT EDIT
+
+#[allow(unused_imports)]
+use libc::{c_int, c_char, c_uchar, c_float, c_uint, c_double,
+    c_short, c_ushort, c_long, c_ulong,
+    c_void, size_t, ssize_t, intptr_t, uintptr_t, time_t, FILE};
+
+#[allow(unused_imports)]
+use glib::{gboolean, gconstpointer, gpointer, GType};
+
+use gobject_subclass::anyimpl::*;
+use gobject_subclass::object::*;
+
+#[cfg(any(feature = "v2_38", feature = "dox"))]
+use Error;
+use ffi;
+use glib;
+use glib::object::Downcast;
+use glib::object::IsA;
+use glib::signal::SignalHandlerId;
+use glib::signal::connect;
+use glib::translate::*;
+use glib_ffi;
+use gobject_ffi;
+use std::boxed::Box as Box_;
+use std::mem;
+use std::mem::transmute;
+use std::ptr;
+
+
+pub trait ActionImpl: AnyImpl + 'static {
+
+    fn activate(&self, action: &T, parameter: Option<&glib::Variant>);
+
+    fn change_state(&self, action: &T, value: &glib::Variant);
+
+    fn get_enabled(&self, action: &T) -> bool;
+
+    fn get_name(&self, action: &T) -> Option<String>;
+
+    fn get_parameter_type(&self, action: &T) -> Option<glib::VariantType>;
+
+    fn get_state(&self, action: &T) -> Option<glib::Variant>;
+
+    fn get_state_hint(&self, action: &T) -> Option<glib::Variant>;
+
+    fn get_state_type(&self, action: &T) -> Option<glib::VariantType>;
+
+}
+
+any_impl!(ActionImpl);
+
+pub trait ActionImplStatic<T: ObjectType>: 'static {
+    fn get_impl<'a>(&self, imp: &'a T::ImplType) -> &'a ActionImpl;
+}
+
+struct ActionStatic<T: ObjectType>{
+    imp_static: *const ActionImplStatic<T>
+}
+
+unsafe extern "C" fn action_activate<T: >
+(ptr: *mut GAction, parameter: *mut glib::GVariant)
+where
+    T::ImplType: ActionImpl<T>
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let klass = &**(ptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(ffi::action_get_type())
+                                     as *const ActionStatic<T>;
+    let interface_static = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    imp.activate()
+}
+
+unsafe extern "C" fn action_change_state<T: >
+(ptr: *mut GAction, value: *mut glib::GVariant)
+where
+    T::ImplType: ActionImpl<T>
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let klass = &**(ptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(ffi::action_get_type())
+                                     as *const ActionStatic<T>;
+    let interface_static = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    imp.change_state()
```

Comment:
Here you want to pass both `ptr` and `value` to the function, which should work the same as in signal trampolines